File: javaLang/QuickSort.java
package javaLang;
import java.io.*;
import java.util.*;

public class QuickSort {
    public static void main(String[] args) throws IOException {
        if (args.length != 1) {
            System.out.println("Usage: java QuickSort <filename>");
            return;
        }

        String filename = args[0];
        List<String> data = new ArrayList<>();

        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                data.add(line.trim());
            }
        }

        long startTime = System.currentTimeMillis();
        quickSort(data, 0, data.size() - 1);
        long endTime = System.currentTimeMillis();

        String outputFilename = "./output/quick_sort_" + data.size() + ".csv";
        new File("./output").mkdirs(); // Create output directory if it doesn't exist
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputFilename))) {
            for (String line : data) {
                writer.write(line);
                writer.newLine();
            }
        }

        System.out.println("Quick Sort completed. Output saved to " + outputFilename);
        System.out.println("Running time: " + (endTime - startTime) + " ms");
    }

    static void quickSort(List<String> arr, int low, int high) {
        // Use a stack to store pairs of (low, high) indices
        Stack<int[]> stack = new Stack<>();
        
        // Push initial bounds onto stack
        stack.push(new int[]{low, high});
        
        // Process stack until empty
        while (!stack.isEmpty()) {
            // Pop low and high from stack
            int[] bounds = stack.pop();
            int currentLow = bounds[0];
            int currentHigh = bounds[1];
            
            // Only partition if low < high
            if (currentLow < currentHigh) {
                // Partition the array and get pivot index
                int pi = partition(arr, currentLow, currentHigh);
                
                // Push left subarray bounds to stack
                if (pi - 1 > currentLow) {
                    stack.push(new int[]{currentLow, pi - 1});
                }
                
                // Push right subarray bounds to stack
                if (pi + 1 < currentHigh) {
                    stack.push(new int[]{pi + 1, currentHigh});
                }
            }
        }
    }

    static int partition(List<String> arr, int low, int high) {
        int pivot = Integer.parseInt(arr.get(high).split(",")[0]);
        int i = low - 1;
        for (int j = low; j < high; j++) {
            int currentVal = Integer.parseInt(arr.get(j).split(",")[0]);
            if (currentVal < pivot) {
                Collections.swap(arr, ++i, j);
            }
        }
        Collections.swap(arr, i + 1, high);
        return i + 1;
    }
}

File: ada_assign/code/pythonLang/quick_sort.py
import sys
import time
import os


def partition(arr, low, high):
    """
    Partition function for QuickSort.
    Places pivot element at its correct position.
    """
    pivot_val = int(arr[high].split(",")[0])
    i = low - 1
    for j in range(low, high):
        curr_val = int(arr[j].split(",")[0])
        if curr_val <= pivot_val:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def quick_sort(arr, low, high):
    """
    Iterative QuickSort implementation using a stack.
    This avoids recursion depth issues.
    """
    # Create a stack to store low and high indices
    stack = []

    # Push initial values onto stack
    stack.append((low, high))

    # Process stack until empty
    while stack:
        # Pop low and high from stack
        low, high = stack.pop()

        # Only partition if low < high
        if low < high:
            # Partition the array and get pivot index
            pi = partition(arr, low, high)

            # Push left subarray bounds to stack
            if pi - 1 > low:
                stack.append((low, pi - 1))

            # Push right subarray bounds to stack
            if pi + 1 < high:
                stack.append((pi + 1, high))


def main():
    if len(sys.argv) != 2:
        print("Usage: python quick_sort.py <filename>")
        return

    filename = sys.argv[1]
    data = []

    try:
        # Read data from file
        with open(filename, "r") as file:
            for line in file:
                line = line.strip()
                if line:
                    data.append(line)

        # Time the sorting operation
        start_time = time.time()
        quick_sort(data, 0, len(data) - 1)
        end_time = time.time()

        running_time = (end_time - start_time) * 1000  # Convert to milliseconds

        # Write sorted data to output file
        output_filename = f"./output/quick_sort_{len(data)}.csv"
        os.makedirs(os.path.dirname(output_filename), exist_ok=True)

        with open(output_filename, "w") as output_file:
            for item in data:
                output_file.write(f"{item}\n")

        print(f"Quick Sort completed. Output saved to {output_filename}")
        print(f"Running time: {running_time:.0f} ms")

    except FileNotFoundError:
        print(f"Error: File '{filename}' not found.")
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()
